src/HydraAuctionOnchain/Lib/Cose.hs

```diff
@@ -0,0 +1,56 @@
+module HydraAuctionOnchain.Lib.Cose
+  ( pmkSigStructure
+  , pverifyCoseSignature
+  ) where
+
+import HydraAuctionOnchain.Lib.Serialization (pserializeAddress)
+import Plutarch.Api.V2 (PAddress)
+import Plutarch.Crypto (pverifyEd25519Signature)
+import Plutarch.Extra.Maybe (pjust, pnothing)
+import Plutarch.Monadic qualified as P
+
+pmkSigStructure
+  :: Term
+      s
+      ( PBool
+          :--> PAddress
+          :--> PByteString
+          :--> PByteString
+          :--> PMaybe PByteString
+      )
+pmkSigStructure = phoistAcyclic $
+  plam $ \isMainnet addr payload payloadLength ->
+    pmatch (pserializeAddress # isMainnet # addr) $ \case
+      PNothing -> pnothing
+      PJust addrSerialized -> P.do
+        addrSerializedFields <- pletFields @["addrCbor", "addrMapEntrySize"] addrSerialized
+        pjust
+          # mconcat
+            [ phexByteStr "846a5369676e61747572653158"
+            , addrSerializedFields.addrMapEntrySize
+            , phexByteStr "A201276761646472657373" -- 61646472657373 = "address"
+            , addrSerializedFields.addrCbor
+            , phexByteStr "4058"
+            , payloadLength
+            , payload
```

Comment:
Does `payloadLength`have a fixed length? Look like it doesn't. Do we need to parse that signature structure?

Reply:
You are correct, the payload length is different for seller and bidder messages.